contracts/votingMachines/GenesisProtocolLogic.sol

```diff
@@ -0,0 +1,660 @@
+        address _voteOnBehalf
+    )
+        public
+        pure
+        returns(bytes32)
+        {
+        //double call to keccak256 to avoid deep stack issue when call with too many params.
+        return keccak256(
+            abi.encodePacked(
+             keccak256(
+              abi.encodePacked(
+                _params[0],
+                _params[1],
+                _params[2],
+                _params[3],
+                _params[4],
+                _params[5],
+                _params[6],
+                _params[7],
+                _params[8],
+                _params[9],
+                _params[10]
+             )),
+            _voteOnBehalf
+        ));
+    }
+
+    /**
+      * @dev execute check if the proposal has been decided, and if so, execute the proposal
+      * @param _proposalId the id of the proposal
+      * @return bool true - the proposal has been executed
+      *              false - otherwise.
+     */
+    function _execute(bytes32 _proposalId) internal votable(_proposalId) returns(bool) {
+        Proposal storage proposal = proposals[_proposalId];
+        Parameters memory params = parameters[proposal.paramsHash];
+        Proposal memory tmpProposal = proposal;
+        uint totalReputation = VotingMachineCallbacksInterface(proposal.callbacks).getTotalReputationSupply(_proposalId);
+        uint executionBar = totalReputation * params.queuedVoteRequiredPercentage/100;
+        ExecutionState executionState = ExecutionState.None;
+        uint averageBoostDownstakes;
+
+        if (proposal.votes[proposal.winningVote] > executionBar) {
+         // someone crossed the absolute vote execution bar.
+            if (proposal.state == ProposalState.Queued) {
+                executionState = ExecutionState.QueueBarCrossed;
+            } else if (proposal.state == ProposalState.PreBoosted) {
+                executionState = ExecutionState.PreBoostedBarCrossed;
+            } else {
+                executionState = ExecutionState.BoostedBarCrossed;
+            }
+            proposal.state = ProposalState.Executed;
+         } else {
+            if (proposal.state == ProposalState.Queued) {
+                // solium-disable-next-line security/no-block-members
+                if ((now - proposal.times[0]) >= params.queuedVotePeriodLimit) {
+                    proposal.state = ProposalState.ExpiredInQueue;
+                    proposal.winningVote = NO;
+                    executionState = ExecutionState.QueueTimeOut;
+                 } else if ( shouldBoost(_proposalId)) {
+                    //change proposal mode to PreBoosted mode.
+                    proposal.state = ProposalState.PreBoosted;
+                    // solium-disable-next-line security/no-block-members
+                    proposal.times[2] = now;
+                    proposal.confidenceThreshold = threshold(proposal.paramsHash,proposal.organizationId);
```

Comment:
at this point we've computed threshold twice (also in shouldBoost), should only compute once.

Reply:
is it possible it might be computed yet again further down in this function?

Reply:
computed yet again only if preBoostedVotePeriodLimit==0 which is a weird configuration..

Reply:
first comment fixed at 699afa8